File: src/main/java/com/github/stars/controller/AiClassifyController.java
package com.github.stars.controller;

import com.github.stars.entity.GithubRepo;
import com.github.stars.service.AiClassifyService;
import com.github.stars.service.GithubRepoService;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import javax.annotation.Resource;
import java.util.*;
import java.util.stream.Collectors;

@Controller
@RequestMapping("/ai")
public class AiClassifyController {

    @Resource
    private AiClassifyService aiClassifyService;

    @Resource
    private GithubRepoService githubRepoService;

    /**
     * AI 分类页面
     */
    @GetMapping("/classify")
    public String classifyPage(
            @RequestParam(value = "language", defaultValue = "") String language,
            @RequestParam(value = "keyword", defaultValue = "") String keyword,
            Model model) {

        List<String> languages = githubRepoService.findAllLanguages();
        model.addAttribute("languages", languages);
        model.addAttribute("language", language);
        model.addAttribute("keyword", keyword);

        return "classify";
    }

    /**
     * 加载仓库列表（用于选择，支持筛选）
     */
    @GetMapping("/classify/repos")
    @ResponseBody
    public Map<String, Object> getRepos(
            @RequestParam(value = "keyword", defaultValue = "") String keyword,
            @RequestParam(value = "language", defaultValue = "") String language) {

        List<GithubRepo> allRepos = githubRepoService.findAll(keyword, language);
        List<Map<String, Object>> repos = allRepos.stream().map(repo -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", repo.getId());
            m.put("repoName", repo.getRepoName());
            m.put("fullName", repo.getFullName());
            m.put("description", repo.getDescription());
            m.put("language", repo.getLanguage());
            m.put("ownerName", repo.getOwnerName());
            m.put("starsCount", repo.getStarsCount());
            m.put("topics", repo.getTopics());
            m.put("htmlUrl", repo.getHtmlUrl());
            return m;
        }).collect(Collectors.toList());

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("repos", repos);
        result.put("total", allRepos.size());
        return result;
    }

    /**
     * 触发 AI 分类
     */
    @PostMapping("/classify/execute")
    @ResponseBody
    public Map<String, Object> executeClassify(@RequestBody Map<String, Object> body) {
        @SuppressWarnings("unchecked")
        List<Integer> repoIdInts = (List<Integer>) body.get("repoIds");
        int topN = body.containsKey("topN") ? ((Number) body.get("topN")).intValue() : 8;

        List<Long> repoIds = repoIdInts.stream().map(Long::valueOf).collect(Collectors.toList());

        Map<String, Object> result = aiClassifyService.classify(repoIds, topN);

        // 为每个分类补充仓库简要信息
        if (result.containsKey("categories")) {
            @SuppressWarnings("unchecked")
            Map<String, List<Long>> categories = (Map<String, List<Long>>) result.get("categories");
            Map<String, List<Map<String, Object>>> enrichedCategories = new LinkedHashMap<>();
            for (Map.Entry<String, List<Long>> entry : categories.entrySet()) {
                List<Map<String, Object>> repoInfos = entry.getValue().stream().map(id -> {
                    GithubRepo repo = githubRepoService.findById(id);
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("id", repo.getId());
                    m.put("repoName", repo.getRepoName());
                    m.put("fullName", repo.getFullName());
                    m.put("description", repo.getDescription());
                    m.put("language", repo.getLanguage());
                    m.put("starsCount", repo.getStarsCount());
                    m.put("htmlUrl", repo.getHtmlUrl());
                    return m;
                }).collect(Collectors.toList());
                enrichedCategories.put(entry.getKey(), repoInfos);
            }
            result.put("categories", enrichedCategories);
        }

        return result;
    }
}

File: backend-nestjs/src/ai/controllers/ai-classify.controller.ts
import { Controller, Get, Post, Body, Query } from '@nestjs/common';
import { AiClassifyService } from '../services/ai-classify.service';
import { GithubRepoService } from '../../github/services/github-repo.service';

@Controller('api')
export class AiClassifyController {
    constructor(
        private readonly classifyService: AiClassifyService,
        private readonly repoService: GithubRepoService,
    ) {}

    /**
     * 获取待分类的仓库列表
     *
     * @param q.keyword 关键词筛选（可选）
     * @param q.language 语言筛选（可选）
     * @returns 仓库列表和总数
     */
    @Get('ai/classify/repos')
    async repos(@Query() q: any) {
        const repos = await this.repoService.findAll({ keyword: q.keyword || '', language: q.language || '' });
        return { repos, total: repos.length };
    }

    /**
     * 执行 AI 普通分类
     *
     * @param b.repoIds 仓库 ID 列表（必填）
     * @param b.topN 最大分类数量（默认 8）
     * @returns 分类结果映射和分类总数
     */
    @Post('ai/classify/execute')
    async execute(@Body() b: any) {
        if (!b.repoIds?.length) return { success: false, message: '请提供仓库ID列表' };
        return this.classifyService.classify(b.repoIds, b.topN || 8);
    }
}
